### ISSUE TITLE
Multiple bounding boxes are detected during visualization

### ISSUE THREAD

**Issue by BENIGA05** [hsiangwei0903/Deep-EIoU#30]
When running demo.py for visualizing yolox models, I'm getting multiple bounding boxes. Can you mention what conf threshold and nms should be passed?

**Comment by hsiangwei0903** [comment]
The default parameter is specified in the file. If you get multiple bounding box for the same person, you can try to use a lower nms threshold.

**Comment by BENIGA05** [comment]
> The default parameter is specified in the file. If you get multiple bounding box for the same person, you can try to use a lower nms threshold.

Adjusted the nms and --conf threshold and checked, still multiple bounding boxes are detected.

**Comment by hsiangwei0903** [comment]
can you provide more details, what threshold did you use? can you also show some visualization results?

**Comment by BENIGA05** [comment]
> can you provide more details, what threshold did you use? can you also show some visualization results?

threshold - 0.1 and nms-0.7

![Image](https://github.com/user-attachments/assets/5026f003-53f0-4207-8421-e2248e0818d0)

**Comment by hsiangwei0903** [comment]
this is beyond threshold issue. what code or script are you running?

**Comment by BENIGA05** [comment]
> this is beyond threshold issue. what code or script are you running?

demo.py for visualization

**Comment by hsiangwei0903** [comment]
this does not look like the results from demo.py? demo.py should contain tracking ID for each target, not just detection results.

**Comment by BENIGA05** [comment]
> this does not look like the results from demo.py? demo.py should contain tracking ID for each target, not just detection results.

I have made changes to demo.py to save detected frames

**Comment by hsiangwei0903** [comment]
likely your changes contain some error? does the tracking results video looks okay?

**Comment by BENIGA05** [comment]
> this does not look like the results from demo.py? demo.py should contain tracking ID for each target, not just detection results.
`import argparse
import os
import os.path as osp
import numpy as np
import time
import cv2
import torch
import sys
from glob import glob
from tqdm import tqdm
sys.path.append('.')
 
from loguru import logger
sys.path.append('/data/users/Beniga/beniga/YOLOX')
from yolox.data.data_augment import preproc
from yolox.exp import get_exp
from yolox.utils import fuse_model, get_model_info, postprocess, vis
# from yolox.utils.visualize import plot_tracking
# from yolox.tracking_utils.timer import Timer
 
# from tracker.Deep_EIoU import Deep_EIoU
# from reid.torchreid.utils import FeatureExtractor
# import torchvision.transforms as T
from yolox.data.datasets import COCO_CLASSES
 
IMAGE_EXT = [".jpg", ".jpeg", ".webp", ".bmp", ".png"]
 
 
def make_parser():
    parser = argparse.ArgumentParser("DeepEIoU Demo")
    parser.add_argument("-expn", "--experiment-name", type=str, default=None)
    parser.add_argument("-n", "--name", type=str, default=None, help="model name")
 
    parser.add_argument(
        "--path", default="/data/users/Beniga/beniga/MOTDataset/Extracted/sportsmot_publish/dataset/sample_test", help="path to images or video"
    )
    parser.add_argument(
        "--save_result",
        default=True,
        help="whether to save the inference result of image/video",
    )
 
    # exp file
    parser.add_argument(
        "-f",
        "--exp_file",
        default="/data/users/Beniga/beniga/Deep-EIoU/Deep-EIoU/yolox/yolox_x_ch_sportsmot.py",
        type=str,
        help="pls input your expriment description file",
    )
    parser.add_argument("-c", "--ckpt", default=None, type=str, help="ckpt for eval")
    parser.add_argument(
        "--device",
        default="gpu",
        type=str,
        help="device to run our model, can either be cpu or gpu",
    )
    parser.add_argument("--conf", default=0.1, type=float, help="test conf")
    parser.add_argument("--nms", default=0.7, type=float, help="test nms threshold")
    parser.add_argument("--tsize", default=None, type=int, help="test img size")
    parser.add_argument("--fps", default=30, type=int, help="frame rate (fps)")
    parser.add_argument(
        "--fp16",
        dest="fp16",
        default=False,
        action="store_true",
        help="Adopting mix precision evaluating.",
    )
    parser.add_argument(
        "--fuse",
        dest="fuse",
        default=False,
        action="store_true",
        help="Fuse conv and bn for testing.",
    )
    parser.add_argument(
        "--trt",
        dest="trt",
        default=False,
        action="store_true",
        help="Using TensorRT model for testing.",
    )
    # tracking args
    parser.add_argument("--track_high_thresh", type=float, default=0.6, help="tracking confidence threshold")
    parser.add_argument("--track_low_thresh", default=0.1, type=float, help="lowest detection threshold valid for tracks")
    parser.add_argument("--new_track_thresh", default=0.7, type=float, help="new track thresh")
    parser.add_argument("--track_buffer", type=int, default=60, help="the frames for keep lost tracks")
    parser.add_argument("--match_thresh", type=float, default=0.8, help="matching threshold for tracking")
    parser.add_argument("--aspect_ratio_thresh", type=float, default=1.6, help="threshold for filtering out boxes of which aspect ratio are above the given value.")
    parser.add_argument('--min_box_area', type=float, default=10, help='filter out tiny boxes')
    parser.add_argument("--nms_thres", type=float, default=0.7,help='nms threshold')
    parser.add_argument("--mot20", dest="mot20", default=False, action="store_true", help="test mot20.")
 
    # reid args
    parser.add_argument("--with-reid", dest="with_reid", default=True, action="store_true", help="use Re-ID flag.")
    parser.add_argument('--proximity_thresh', type=float, default=0.5, help='threshold for rejecting low overlap reid matches')
    parser.add_argument('--appearance_thresh', type=float, default=0.25, help='threshold for rejecting low appearance similarity reid matches')
    return parser
 
class Predictor(object):
    def __init__(
        self,
        model,
        exp,
        trt_file=None,
        decoder=None,
        device=torch.device("cpu"),
        fp16=False
    ):
        self.model = model
        self.decoder = decoder
        self.num_classes = exp.num_classes
        self.confthre = exp.test_conf
        print(f"conf threshold {self.confthre}")
        self.nmsthre = exp.nmsthre
        self.test_size = exp.test_size
        self.device = device
        self.fp16 = fp16
        self.cls_names = COCO_CLASSES  
        if trt_file is not None:
            from torch2trt import TRTModule
 
            model_trt = TRTModule()
            model_trt.load_state_dict(torch.load(trt_file))
 
            x = torch.ones((1, 3, exp.test_size[0], exp.test_size[1]), device=device)
            self.model(x)
            self.model = model_trt
        self.rgb_means = (0.485, 0.456, 0.406)
        self.std = (0.229, 0.224, 0.225)
 
    def inference(self, img):
        img_info = {"id": 0}
        if isinstance(img, str):
            img_info["file_name"] = osp.basename(img)
            img = cv2.imread(img)
        else:
            img_info["file_name"] = None
 
        height, width = img.shape[:2]
        img_info["height"] = height
        img_info["width"] = width
        img_info["raw_img"] = img
 
        img, ratio = preproc(img, self.test_size)
        img_info["ratio"] = ratio
        img = torch.from_numpy(img).unsqueeze(0).float().to(self.device)
        if self.fp16:
            img = img.half()  # to FP16
 
        with torch.no_grad():
            # timer.tic()
            outputs = self.model(img)
            if self.decoder is not None:
                outputs = self.decoder(outputs, dtype=outputs.type())
            outputs = postprocess(
                outputs, self.num_classes, self.confthre, self.nmsthre, class_agnostic=False
            )
            for i in range(len(outputs)):
                outputs[i] = outputs[i].cpu().detach().numpy().tolist()
       
        return outputs, img_info
    def visual(self, output, img_info, cls_conf=0.35):
        ratio = img_info["ratio"]
        img = img_info["raw_img"]
        if output is None:
            return img
        output = torch.tensor(output[0])
 
        bboxes = output[:, 0:4]
 
        # preprocessing: resize
        bboxes /= ratio
 
        cls = output[:, 6]
        scores = output[:, 4] * output[:, 5]
 
        vis_res = vis(img, bboxes, scores, cls, cls_conf, self.cls_names)
        return vis_res
    
 
 
def main(exp, args):
    if not args.experiment_name:
        args.experiment_name = exp.exp_name
 
    output_dir = osp.join(exp.output_dir, args.experiment_name)
    os.makedirs(output_dir, exist_ok=True)
 
    vis_folder = osp.join(output_dir, "track_vis")
    os.makedirs(vis_folder, exist_ok=True)
 
    if args.trt:
        args.device = "gpu"
    args.device = torch.device("cuda" if args.device == "gpu" else "cpu")
 
    logger.info("Args: {}".format(args))
 
    if args.conf is not None:
        exp.test_conf = args.conf
    if args.nms is not None:
        exp.nmsthre = args.nms
    if args.tsize is not None:
        exp.test_size = (args.tsize, args.tsize)
 
    model = exp.get_model().to(args.device)
    logger.info("Model Summary: {}".format(get_model_info(model, exp.test_size)))
    model.eval()
 
    if not args.trt:
        if args.ckpt is None:
            ckpt_file = "/data/users/Beniga/beniga/YOLOX/yolox/best_ckpt.pth.tar"
        else:
            ckpt_file = args.ckpt
        logger.info("loading checkpoint")
        ckpt = torch.load(ckpt_file, map_location="cpu")
        # load the model state dict
        model.load_state_dict(ckpt["model"])
        logger.info("loaded checkpoint done.")
 
    if args.fuse:
        logger.info("\tFusing model...")
        model = fuse_model(model)
 
    if args.fp16:
        model = model.half()  # to FP16
 
    if args.trt:
        assert not args.fuse, "TensorRT model is not support model fusing!"
        trt_file = osp.join(output_dir, "model_trt.pth")
        assert osp.exists(
            trt_file
        ), "TensorRT model is not found!\n Run python3 tools/trt.py first!"
        model.head.decode_in_inference = False
        decoder = model.head.decode_outputs
        logger.info("Using TensorRT to inference")
    else:
        trt_file = None
        decoder = None
 
    predictor = Predictor(model, exp, trt_file, decoder, args.device, args.fp16)
    current_time = time.localtime()
    print("Inside")
 
    output_dir = "/data/users/Beniga/beniga/Deep-EIoU/detection"
    sequence_dirs = sorted(glob(os.path.join(
    "/data/users/Beniga/beniga/MOTDataset/Extracted/sportsmot_publish/dataset/sample_test", "*", "img1")))
 
    # sequence_dirs = sorted(glob("/data/users/Beniga/beniga/sn-tracking/test1*/**/img1", recursive=True))
 
 
    for img1_path in sequence_dirs:
        print("Inside")
        sequence_name = os.path.basename(os.path.dirname(img1_path))
        output_npy_path = os.path.join(output_dir, f"{sequence_name}.npy")
        img_paths = sorted([
            os.path.join(img1_path, f) for f in os.listdir(img1_path)
            if f.endswith(('.jpg', '.png'))
        ])
        all_dets = []
 
        # Set up save dir for each sequence
        save_dir = os.path.abspath(os.path.join("../", "output_img", sequence_name, "img1"))
        os.makedirs(save_dir, exist_ok=True)
 
        for frame_idx, image_path in enumerate(tqdm(img_paths, desc=f"Processing {sequence_name}")):
            outputs, img_info = predictor.inference(image_path)
 
            result_image = predictor.visual(outputs, img_info, predictor.confthre)
 
            save_file_name = os.path.join(save_dir, f"frame_{frame_idx:04d}.jpg")
            success = cv2.imwrite(save_file_name, result_image)
            if not success:
                logger.warning(f"❌ Failed to save image: {save_file_name}")
            else:
                logger.info(f"✅ Saved detection result: {save_file_name}")
 
            all_dets.append(outputs)
 
        # Save detections as .npy
        all_dets = [[np.array(det, dtype=np.float32) for det in frame] for frame in all_dets]
        np.save(output_npy_path, np.array(all_dets, dtype=object))
        print(f"✅ Saved detections to {output_npy_path}")
 
if __name__ == "__main__":
    args = make_parser().parse_args()
    exp = get_exp(args.exp_file, args.name)
 
    main(exp, args)
 
 `

**Comment by BENIGA05** [comment]
> likely your changes contain some error? does the tracking results video looks okay?

no video is not saved correctly as per the code. So made some changes to save detected images

**Comment by hsiangwei0903** [comment]
demo.py should save the tracking results as a video, what error did you have when running demo.py?

**Comment by BENIGA05** [comment]
> demo.py should save the tracking results as a video, what error did you have when running demo.py?

the print message is printed whereas the video is not saved in that path. Does any changes to be done to the code to get the result?

**Comment by BENIGA05** [comment]
> demo.py should save the tracking results as a video, what error did you have when running demo.py?

2025-07-08 00:15:18.888 | INFO     | __main__:main:613 - Args: Namespace(experiment_name='yolox_x_ch_sportsmot', name=None, path='/data/users/Beniga/beniga/MOTDataset/Extracted/sportsmot_publish/dataset/sample_test', save_result=True, exp_file='/data/users/Beniga/beniga/Deep-EIoU/Deep-EIoU/yolox/yolox_x_ch_sportsmot.py', ckpt=None, device=device(type='cuda'), conf=0.1, nms=0.7, tsize=None, fps=30, fp16=False, fuse=False, trt=False, track_high_thresh=0.6, track_low_thresh=0.1, new_track_thresh=0.7, track_buffer=60, match_thresh=0.8, aspect_ratio_thresh=1.6, min_box_area=10, nms_thres=0.7, mot20=False, with_reid=True, proximity_thresh=0.5, appearance_thresh=0.25)
2025-07-08 00:15:19.707 | INFO     | __main__:main:623 - Model Summary: Params: 99.00M, Gflops: 793.21
2025-07-08 00:15:19.708 | INFO     | __main__:main:631 - loading checkpoint
2025-07-08 00:15:20.079 | INFO     | __main__:main:635 - loaded checkpoint done.
/data/users/Beniga/beniga/Deep-EIoU/Deep-EIoU/reid/torchreid/utils/torchtools.py:297: UserWarning: The pretrained weights "/data/users/Beniga/beniga/gta-link/reid_checkpoints/best_ckpt.pth.tar" cannot be loaded, please check the key names manually (** ignored and continue **)
  warnings.warn(
2025-07-08 00:15:20.489 | INFO     | __main__:imageflow_demo:474 - video save_path is ./YOLOX_outputs/yolox_x_ch_sportsmot/track_vis/2025_07_08_00_15_20/sample_test
[ WARN:0@2.847] global cap.cpp:779 open VIDEOIO(CV_IMAGES): raised OpenCV exception:

OpenCV(4.11.0) /io/opencv/modules/videoio/src/cap_images.cpp:415: error: (-215:Assertion failed) !filename_pattern.empty() in function 'CvVideoWriter_Images'


2025-07-08 00:15:20.544 | INFO     | __main__:imageflow_demo:532 - Processing frame 0 (100000.00 fps)
2025-07-08 00:15:20.558 | INFO     | __main__:imageflow_demo:590 - ✅ Video saved and finalized at: ./YOLOX_outputs/yolox_x_ch_sportsmot/track_vis/2025_07_08_00_15_20/sample_test
2025-07-08 00:15:20.573 | INFO     | __main__:imageflow_demo:596 - save results to ./YOLOX_outputs/yolox_x_ch_sportsmot/track_vis/2025_07_08_00_15_20.txt



This is the output I m getting while running demo.py. Here only the print is printed and the file is not saved in that path

**Comment by hsiangwei0903** [comment]
i think you did not successfully load the reid model, the video is also not read correctly.
perhaps try to solve these issues first instead of directly modifying the code in demo.py

**Comment by BENIGA05** [comment]
> i think you did not successfully load the reid model, the video is also not read correctly. perhaps try to solve these issues first instead of directly modifying the code in demo.py

Now with this same code, I got a detected video result, in which the bounding boxes are not aligned properly to the player

**Comment by hsiangwei0903** [comment]
Probably the bounding box format is not correct when plotting the video. Try to check if the format is xyxy or xywh.

**Comment by BENIGA05** [comment]
> Probably the bounding box format is not correct when plotting the video. Try to check if the format is xyxy or xywh.

the format is xywh

**Comment by hsiangwei0903** [comment]
I am closing this issue due to inactivity, feel free to reopen if any question.
